Disable CXX 11 ABI when building wheels

jysohn23 opened the pull request:
Also for CircleCI checks.

jysohn23 commented:
> If we do this (exporting variables) in our `setup.py`, we should cover all cases automatically.
> 
> Also, this (defining that macro) is kinda of a shotgun approach IMHO. PT code should not be defining ABI tags if they then set that macro to 0 during builds.

PT sets `D_GLIBCXX_USE_CXX11_ABI=0` during their release builds to be compatible with RHEL7 apparently, whereas in their CI they don't. I agree it'd be nice if its consistent, but I think this works for us for the time being.

And good idea on the `setup.py` I'll set the env var for clang to pickup there, though I *think* bazel needs that passed in as a `-copt` (correct me if I'm wrong).

jysohn23 commented:
@davidel I tried setting it in `xla/setup.py` but that still compiles with `-D_GLIBCXX_USE_CXX11_ABI=1`. It actually shows both `-D_GLIBCXX_USE_CXX11_ABI=1` and `-D_GLIBCXX_USE_CXX11_ABI=0`:

```
clang-8 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -D_GLIBCXX_USE_CXX11_ABI=0 ... -DTORCH_EXTENSION_NAME=_XLAC -D_GLIBCXX_USE_CXX11_ABI=1
```

This is most likely due to `BuildExtension` setting up this `-D_GLIBCXX_USE_CXX11_ABI=1` itself from here: https://github.com/pytorch/pytorch/blob/master/torch/utils/cpp_extension.py#L704

> If we do this (exporting variables) in our setup.py, we should cover all cases automatically.

If only setting `=0` from `xla/setup.py`, then `pytorch/setup.py` would still be built with `-D_GLIBCXX_USE_CXX11_ABI=1` and thus `torch._C._GLIBCXX_USE_CXX11_ABI=1`. Building `torch` with `=1` and trying to build `torch_xla` with `=0` I don't think makes sense and wouldn't work (which is what we'd be doing)?

davidel commented:
Good thing they export it, but our setup in `build_torch_xla_libs.sh` should come from `setup.py` reading it.

```diff
diff --git a/build_torch_xla_libs.sh b/build_torch_xla_libs.sh
index eea48c80..cf0bc8bb 100755
--- a/build_torch_xla_libs.sh
+++ b/build_torch_xla_libs.sh
@@ -2,6 +2,20 @@
 
 set -ex
 
+OPTS=()
+
+while getopts "O:" OPTION
+do
+    case $OPTION in
+       O)
+           for i in ${OPTARG}; do
+               OPTS+=("--cxxopt=${i}")
+           done
+           ;;
+    esac
+done
+shift $(($OPTIND - 1))
+
 CMD="${1:-install}"
 
 cd "$(dirname "$0")"
@@ -25,7 +39,7 @@ if [[ "$XLA_CUDA" == "1" ]] && [[ "$CLOUD_BUILD" == "true" ]]; then
   MAX_JOBS="--jobs=16"
 fi
 
-OPTS=(--cxxopt="-std=c++14")
+OPTS+=(--cxxopt="-std=c++14")
 if [[ $(basename -- $CC) =~ ^clang ]]; then
   OPTS+=(--cxxopt="-Wno-c++11-narrowing")
 fi
diff --git a/setup.py b/setup.py
index 8d7d65b2..297a5c7e 100644
--- a/setup.py
+++ b/setup.py
@@ -16,6 +16,7 @@ import re
 import shutil
 import subprocess
 import sys
+import torch
 
 base_dir = os.path.dirname(os.path.abspath(__file__))
 third_party_path = os.path.join(base_dir, 'third_party')
@@ -86,6 +87,9 @@ def generate_xla_aten_code(base_dir):
 
 def build_extra_libraries(base_dir, build_mode=None):
   build_libs_cmd = [os.path.join(base_dir, 'build_torch_xla_libs.sh')]
+  cxx_abi = getattr(torch._C, '_GLIBCXX_USE_CXX11_ABI', None)
+  if cxx_abi is not None:
+    build_libs_cmd += ['-O', '-D_GLIBCXX_USE_CXX11_ABI={}'.format(int(cxx_abi))]
   if build_mode is not None:
     build_libs_cmd += [build_mode]
   if subprocess.call(build_libs_cmd) != 0:
```

davidel commented on scripts/build_torch_wheels.sh:
```diff
@@ -211,9 +211,14 @@ function main() {
   build_and_install_torch
   pushd xla
   build_and_install_torch_xla
+
```
Why this line break?
If you really want to change the format, maybe something like this?
```shell
something

pushd FOO
foo_cmd ...
popd

something_else
```

jysohn23 replied:
Undone.

davidel commented on scripts/build_torch_wheels.sh:
```diff
@@ -211,9 +211,14 @@ function main() {
+
+  # Upstream PyTorch is released with CXX11 ABI disabled
+  # so to avoid symbol mismatch errors we must do the same.
+  export CFLAGS="${CFLAGS} -D_GLIBCXX_USE_CXX11_ABI=0" CXXFLAGS="${CXXFLAGS} -D_GLIBCXX_USE_CXX11_ABI=0"
```
I think we can avoid bolting this here, and have the caller (our build machinery that builds wheels for Colab/Kaggle) export CFLAGS.

jysohn23 replied:
Done.

davidel commented on build_torch_xla_libs.sh:
```diff
@@ -43,8 +43,8 @@ else
   pushd $THIRD_PARTY_DIR/tensorflow
-  bazel build $MAX_JOBS $VERBOSE --define framework_shared_object=false -c "$MODE" "${OPTS[@]}" $XLA_CUDA_CFG \
-    //tensorflow/compiler/xla/xla_client:libxla_computation_client.so
+  bazel build $MAX_JOBS $VERBOSE --define framework_shared_object=false -c "$MODE" --copt="-D_GLIBCXX_USE_CXX11_ABI=0" \
```
Check the diff I sent, WRT this coming from torch._C -> setup.py ...

jysohn23 replied:
Done, thanks!

davidel commented:
I’m out ATM but if you addressed the comments ask Ailing to review and go
out in!


On Mon, Aug 3, 2020 at 21:05 Jin Young Sohn <notifications@github.com>
wrote:

> @jysohn23 <https://github.com/jysohn23> requested your review on: #2390
> <https://github.com/pytorch/xla/pull/2390> Disable CXX 11 ABI when
> building wheels.
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/xla/pull/2390#event-3617224086>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALVJPEYBKQJSADUKULMVGLR64C7ZANCNFSM4PSKHFYA>
> .
>

jysohn23 commented:
> I’m out ATM but if you addressed the comments ask Ailing to review and go out in!
> […](#)
> On Mon, Aug 3, 2020 at 21:05 Jin Young Sohn ***@***.***> wrote: @jysohn23 <[@jysohn23](https://github.com/jysohn23)> requested your review on: #2390 <#2390> Disable CXX 11 ABI when building wheels. — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub <[#2390 (comment)](https://github.com/pytorch/xla/pull/2390#event-3617224086)>, or unsubscribe <[github.com/notifications/unsubscribe-auth/AALVJPEYBKQJSADUKULMVGLR64C7ZANCNFSM4PSKHFYA](https://github.com/notifications/unsubscribe-auth/AALVJPEYBKQJSADUKULMVGLR64C7ZANCNFSM4PSKHFYA)> .

Will do thanks!
